libs/core/langchain_core/indexes/base.py

```diff
@@ -0,0 +1,369 @@
+import warnings
+from abc import ABC, abstractmethod
+from typing import (
+    Any,
+    AsyncIterable,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+    cast,
+)
+
+from langchain_core.documents import Document
+from langchain_core.indexes.types import DeleteResponse, UpsertResponse
+from langchain_core.runnables import run_in_executor
+from langchain_core.stores import BaseStore
+from langchain_core.structured_query import StructuredQuery
+
+
+class Index(BaseStore[str, Document], ABC):
```

Comment:
Should we add any ability to add unique hash to identify a document or information about time of modification / creation